js/modules.go

```diff
@@ -0,0 +1,122 @@
+package js
+
+import (
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/dop251/goja"
+	"go.k6.io/k6/js/compiler"
+	"go.k6.io/k6/js/modules"
+	"go.k6.io/k6/loader"
+)
+
+type module interface {
+	Instantiate(vu modules.VU) moduleInstance
+}
+
+type moduleInstance interface {
+	execute() error
+	exports() *goja.Object
+}
+type moduleCacheElement struct {
+	mod module
+	err error
+}
+
+type moduleResolver struct {
+	cache     map[string]moduleCacheElement
+	goModules map[string]interface{}
+}
+
+func newModuleResolution(goModules map[string]interface{}) *moduleResolver {
+	return &moduleResolver{goModules: goModules, cache: make(map[string]moduleCacheElement)}
+}
+
+func (mr *moduleResolver) setMain(main *loader.SourceData, c *compiler.Compiler) error {
+	mod, err := cjsmoduleFromString(main.URL, main.Data, c)
+	mr.cache[main.URL.String()] = moduleCacheElement{mod: mod, err: err}
+	return err
+}
+
+func (mr *moduleResolver) resolveSpecifier(basePWD *url.URL, arg string) (*url.URL, error) {
+	specifier, err := loader.Resolve(basePWD, arg)
+	if err != nil {
+		return nil, err
+	}
+	return specifier, nil
+}
+
+func (mr *moduleResolver) requireModule(name string) (module, error) {
+	mod, ok := mr.goModules[name]
+	if !ok {
+		return nil, fmt.Errorf("unknown module: %s", name)
+	}
+	if m, ok := mod.(modules.Module); ok {
+		return &goModule{Module: m}, nil
+	}
+
+	return &baseGoModule{mod: mod}, nil
+}
+
+func (mr *moduleResolver) resolve(basePWD *url.URL, arg string, loadCJS cjsModuleLoader) (module, error) {
+	if cached, ok := mr.cache[arg]; ok {
+		return cached.mod, cached.err
+	}
+	switch {
+	case arg == "k6", strings.HasPrefix(arg, "k6/"):
+		// Builtin or external modules ("k6", "k6/*", or "k6/x/*") are handled
+		// specially, as they don't exist on the filesystem.
+		mod, err := mr.requireModule(arg)
+		mr.cache[arg] = moduleCacheElement{mod: mod, err: err}
+		return mod, err
+	default:
+		specifier, err := mr.resolveSpecifier(basePWD, arg)
+		if err != nil {
+			return nil, err
+		}
+		// try cache with the final specifier
+		if cached, ok := mr.cache[specifier.String()]; ok {
+			return cached.mod, cached.err
+		}
+		// Fall back to loading from the filesystem.
+		mod, err := loadCJS(specifier, arg)
+		mr.cache[specifier.String()] = moduleCacheElement{mod: mod, err: err}
+		return mod, err
+	}
+}
+
+type moduleSystem struct {
+	vu            modules.VU
+	instanceCache map[module]moduleInstance
+	resolver      *moduleResolver
+	cjsLoad       cjsModuleLoader
+}
+
+func newModuleSystem(resolution *moduleResolver, vu modules.VU, cjsLoad cjsModuleLoader) *moduleSystem {
```

Comment:
Shouldn't the `cjsModuleLoader` be part of the `moduleResolver` and not the `moduleSystem`? :thinking: 

You are not using it here, you are just saving it in the object and passing it as an argument to `ms.resolver.resolve()` on every call. It will be simpler to just pass it once to `newModuleResolution()`, right?

Reply:
Fixed in [f020530](https://github.com/grafana/k6/pull/2881/commits/f0205307e0135b48e84bcb08b497776c3fbd738a) 

I am pretty sure this was the way easier thing to do before some other refactor, but now this is definitely an improvement :tada: 
good catch :clap: